metadata/src/main/java/org/apache/kafka/image/ClientQuotasImage.java

```diff
@@ -126,40 +144,45 @@ public DescribeClientQuotasResponseData describe(DescribeClientQuotasRequestData
                     "user or clientId filter component.");
             }
         }
-        // TODO: this is O(N). We should add indexing here to speed it up. See KAFKA-13022.
-        for (Entry<ClientQuotaEntity, ClientQuotaImage> entry : entities.entrySet()) {
-            ClientQuotaEntity entity = entry.getKey();
-            ClientQuotaImage quotaImage = entry.getValue();
-            if (matches(entity, exactMatch, typeMatch, request.strict())) {
-                response.entries().add(toDescribeEntry(entity, quotaImage));
-            }
-        }
-        return response;
-    }
 
-    private static boolean matches(ClientQuotaEntity entity,
-                                   Map<String, String> exactMatch,
-                                   Set<String> typeMatch,
-                                   boolean strict) {
-        if (strict) {
-            if (entity.entries().size() != exactMatch.size() + typeMatch.size()) {
-                return false;
+        Set<ClientQuotaEntity> addedEntities = new HashSet<>();
+        for (Entry<String, String> exactMatchEntry : exactMatch.entrySet()) {
+            if (entitiesByType.containsKey(exactMatchEntry.getKey()) &&
+                entitiesByType.get(exactMatchEntry.getKey()).containsKey(exactMatchEntry.getValue())) {
+                for (Entry<ClientQuotaEntity, ClientQuotaImage> entry : entitiesByType.get(exactMatchEntry.getKey()).get(exactMatchEntry.getValue())) {
+                    if (request.strict() && !entry.getKey().entries().equals(exactMatch)) {
```

Comment:
does this have to equal exactMatch as long as entry.getKey().entries() are encompassed by both typeMatch and exactMatch?